agent/src/beerocks/slave/agent_db.h

```diff
@@ -0,0 +1,76 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2019-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "task_pool.h"
+
+#include <iostream>
+#include <memory>
+#include <mutex>
+
+using namespace beerocks_message;
+using namespace son;
+using namespace beerocks;
+
+/**
+ * @brief AgentDB is a class used by all main Agent threads on the Agent process to store the Agent
+ * common data.
+ * The class is implemented using singleton design pattern.
+ * It is thread safe and being locked and released automatically.
+ * 
+ * How to use:
+ * 
+ * auto db = AgentDB::get();            // Get DB and automatically lock it. The lock will be
+ *                                      // Release when 'db' will be destructed.
+ * 
+ * db->foo = 42;                        // Change database member 'foo' to value 42
+ * db->bar = std::string("prpl")        // Set another database member
+ * 
+ * AgentDB::get()->foo = 44;            // Set database member directly.
+ * auto my_foo = AgentDB::get()->foo;   // Get a copy of database member.
+ * 
+ * auto &foo = AgentDB::get()->foo;     // Unsafe! Don't do it! The database will be unlocked,
+ *                                      // but the reference will remain.
+ */
+class AgentDB {
+public:
+    class SafeDB {
+    public:
+        SafeDB(AgentDB &db) : m_db(db) { m_db.db_lock(); }
+        ~SafeDB() { m_db.db_unlock(); }
+        AgentDB *operator->() { return &m_db; }
+
+    private:
+        AgentDB &m_db;
+    };
+    static SafeDB get()
+    {
+        // Guaranteed to be destroyed.
+        // Instantiated on first use.
+        static AgentDB instance;
+        return SafeDB(instance);
+    }
+    AgentDB(AgentDB const &) = delete;
+    void operator=(AgentDB const &) = delete;
+
+private:
+    // Private constructor so that no objects can be created.
+    AgentDB() {}
+    std::recursive_mutex m_db_mutex;
+    void db_lock() { m_db_mutex.lock(); }
+    void db_unlock() { m_db_mutex.unlock(); }
+
+public:
+    /* Put here database members used by the Agent modules */
+
+    /* Tasks */
+    task_pool tasks;
+
+    struct sTasksId {
+        int topology_task = -1;
+    } tasks_id;
```

Comment:
So, this would be an example of using identifiers rather than references. Like this, it's pretty useless, you would need something to get the task object based on the task ID as well.

Since it's not used yet, I would just remove the tasks and tasks_id for the moment - it makes much more sense to add them once you start using them.

Reply:
removed